### PR TITLE
[ibmcloud] introduce provision resource group step and relief usage of the same base domain

### PR DESCRIPTION
**Pull request by jianlinliu**
1. Because IBMCloud dns service has "the maximum limit of 5 wildcard records" (IBMCloud dns service is only used in private cluster install, non-private cluster install uses CIS dns), if running more than 5 concurrent private clusters using the same base domain, the install would fail on ingress operator can not provision wildcard apps dns. So create several zones with different base domain to relief the usage of the same base domain.
```
ibmcloud.qe.devcluster.openshift.com -> disconnected private cluster install in prow ci
private-ibmcloud.qe.devcluster.openshift.com -> all private cluster install for flexy templates
private-ibmcloud-1.qe.devcluster.openshift.com -> basic private cluster install in prow ci
private-ibmcloud-2.qe.devcluster.openshift.com -> proxy private cluster install in prow ci
```
2. Decrease some job's test frequency using `ibmcloud.qe.devcluster.openshift.com`
3. Merge some install configurations (kms-key and with-ep-cos) into private cluster to relief the usage of `ibmcloud.qe.devcluster.openshift.com`
4. When merging some install configurations into private cluster, e.g: kms + private, found it was creating resource group with the same name twice, one is in vpc provision, another is in kms provision, so extract the step of resource group step into a dedicated step, so that we can easily opt-in the step when needed.  This change affects many install workflow, e.g: kms-key, vpc, bastionhost, static dedicated host.

### CODE REVIEW

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-ibmcloud-ipi-private-custom-cos-endpoint-fips-f28 periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-ibmcloud-ipi-disc-priv-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

**Comment by jianlinliu** [comment]
cc @MayXuQQ to review

**Comment by MayXuQQ** [comment]
could you help rerun ibmcloud-ipi-dedicatedhost-static ?

**Comment by jianlinliu** [comment]
/pj-rehearse ci-openshift-openshift-tests-private-release-4.15-amd64-stable-ibmcloud-ipi-disc-priv-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-ibmcloud-ipi-disc-priv-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-ibmcloud-ipi-dedicatedhost-static-f28

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-amd64-stable-ibmcloud-ipi-private-fips-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.12-amd64-stable-ibmcloud-ipi-private-fips-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.12-amd64-stable-ibmcloud-ipi-private-fips-f14

File: ci-operator/step-registry/ibmcloud/deprovision/resourcegroup/ibmcloud-deprovision-resourcegroup-commands.sh [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      return 2
+  fi
+  echo "$output"
+  return 0
+}
+
+ibmcloud_login
+
+rg_files="${SHARED_DIR}/ibmcloud_resource_group"
+for rg_file in ${rg_files}; do
```
Comment (MayXuQQ) [inline ci-operator/step-registry/ibmcloud/deprovision/resourcegroup/ibmcloud-deprovision-resourcegroup-commands.sh:53]: not just one file named "${SHARED_DIR}/ibmcloud_resource_group", saved the resource group name?

File: ci-operator/step-registry/ibmcloud/provision/resourcegroup/ibmcloud-provision-resourcegroup-commands.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+function create_resource_group() {
+    local rg="$1"
+    echo "create resource group ... ${rg}"
+    "${IBMCLOUD_CLI}" resource group-create ${rg} || return 1
+    "${IBMCLOUD_CLI}" target -g ${rg} || return 1
+}
+
+ibmcloud_login
+
+rg_name="${NAMESPACE}-${UNIQUE_HASH}"
```
Comment (MayXuQQ) [inline ci-operator/step-registry/ibmcloud/provision/resourcegroup/ibmcloud-provision-resourcegroup-commands.sh:27]: how about adding '-rg', to distinguish the resource group which auto-created?

File: ci-operator/step-registry/ibmcloud/provision/vpc/ibmcloud-provision-vpc-commands.sh [hunk 3]
```diff
@@ -167,11 +208,16 @@ workdir="$(mktemp -d)"
...
+else
+    rg_name_line="networkResourceGroupName: ${resource_group}"
+fi
+
 cat > "${SHARED_DIR}/customer_vpc_subnets.yaml" << EOF
 platform:
   ibmcloud:
-    resourceGroupName: ${resource_group}
-    networkResourceGroupName: ${resource_group}
+    ${rg_name_line}
```
Comment (MayXuQQ) [inline ci-operator/step-registry/ibmcloud/provision/vpc/ibmcloud-provision-vpc-commands.sh:220]: shall we need cover resouceGroupName is not same as networkResourceGroupName?

File: ci-operator/step-registry/ibmcloud/deprovision/resourcegroup/ibmcloud-deprovision-resourcegroup-commands.sh (hunk 1 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ibmcloud/deprovision/resourcegroup/ibmcloud-deprovision-resourcegroup-commands.sh:53]: in the future, we can easily extend the file list, so use for loop here.

File: ci-operator/step-registry/ibmcloud/provision/vpc/ibmcloud-provision-vpc-commands.sh (hunk 3 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ibmcloud/provision/vpc/ibmcloud-provision-vpc-commands.sh:220]: That depends on your test case, that is out of this PR's scope. Whatever, I introduced `CREATE_CLUSTER_RESOURCE_GROUP` in `ibmcloud-provision-resourcegroup` to support that behavior.

File: ci-operator/step-registry/ibmcloud/provision/resourcegroup/ibmcloud-provision-resourcegroup-commands.sh (hunk 2 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ibmcloud/provision/resourcegroup/ibmcloud-provision-resourcegroup-commands.sh]: fixed.

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

File: ci-operator/step-registry/ibmcloud/deprovision/resourcegroup/ibmcloud-deprovision-resourcegroup-commands.sh (hunk 1 above)
Comment (MayXuQQ) [reply ci-operator/step-registry/ibmcloud/deprovision/resourcegroup/ibmcloud-deprovision-resourcegroup-commands.sh:53]: now have not a step which creates file ${SHARED_DIR}/ibmcloud_cluster_resource_group? 
how about hiding the code which may extend in the future? in order to be easy to understand?

Comment (jianlinliu) [reply ci-operator/step-registry/ibmcloud/deprovision/resourcegroup/ibmcloud-deprovision-resourcegroup-commands.sh:53]: > now have not a step which creates file ${SHARED_DIR}/ibmcloud_cluster_resource_group?

No, I never saw that.

> how about hiding the code which may extend in the future? in order to be easy to understand?

I already added it when resolving https://github.com/openshift/release/pull/49541#discussion_r1514751928.

**Comment by MayXuQQ** [comment]
could you help remove 'set -x' in ci-operator/step-registry/cucushift/installer/check/ibmcloud/destroy/cucushift-installer-check-ibmcloud-destroy-commands.sh L24? thanks a lot

**Comment by jianlinliu** [comment]
> could you help remove 'set -x' in ci-operator/step-registry/cucushift/installer/check/ibmcloud/destroy/cucushift-installer-check-ibmcloud-destroy-commands.sh L24? thanks a lot

I remember you want all the ibmcloud install workflow to run the destroy check. That is why you added it, right?

**Comment by MayXuQQ** [comment]
> > could you help remove 'set -x' in ci-operator/step-registry/cucushift/installer/check/ibmcloud/destroy/cucushift-installer-check-ibmcloud-destroy-commands.sh L24? thanks a lot
> 
> I remember you want all the ibmcloud install workflow to run the destroy check. That is why you added it, right?

just remove 'set -x' in L24

**Comment by jianlinliu** [comment]
> just remove 'set -x' in L24

Got it, fixed.

**Comment by jianlinliu** [comment]
The PR is ready to merge, if it looks good to you, pls drop `/lgtm` there. Thanks.

**Comment by jianlinliu** [comment]
cc @liangxia to review.

**Pull request merged**